Graphics/MaliCompilerReport/Assets/Editor/MaliShaderReport.cs

```diff
@@ -0,0 +1,644 @@
+
+            StringBuilder sb = new StringBuilder(k_MaxHintLength * 3);
+
+            for (int sub = 0; sub < data.SubshaderCount; sub++)
+            {
+                ShaderData.Subshader subshader = data.GetSubshader(sub);
+                for (int p = 0; p < subshader.PassCount; p++)
+                {
+                    PassInfo passInfo = new PassInfo();
+                    passInfo.subShader = sub;
+                    passInfo.pass = subshader.GetPass(p);
+                    passInfo.name = passInfo.pass.Name;
+
+                    ParseKeywords(passInfo.pass.SourceCode, ref passInfo.keywords);
+
+                    // Create some label hints
+                    for (int i = 0; i < passInfo.keywords.Count; i++)
+                    {
+                        for (int j = 0; j < passInfo.keywords[i].Count; j++)
+                        {
+                            if (passInfo.keywords[i][j].Length > 0)
+                            {
+                                sb.Append(passInfo.keywords[i][j]);
+
+                                if (sb.Length > k_MaxHintLength)
+                                {
+                                    sb.Remove(k_MaxHintLength, sb.Length - k_MaxHintLength);
+                                    sb[15] = '.';
+                                    sb[16] = '.';
+                                    sb[17] = '.';
+
+                                    break;
+                                }
+                                else if (sb.Length == k_MaxHintLength)
+                                {
+                                    break;
+                                }
+                            }
+                        }
+
+                        passInfo.keywordHint.Add(sb.ToString());
+                        sb.Clear();
+                    }
+
+                    m_Passes.Add(passInfo);
+                    names.Add(passInfo.name);
+                }
+            }
+
+            if (m_SelectedPass < 0)
+                m_SelectedPass = 0;
+            else if (m_SelectedPass >= m_Passes.Count)
+            {
+                m_SelectedPass = m_Passes.Count - 1;
+            }
+
+            SetSelectedPass(m_SelectedPass, true);
+
+            m_PassNames = names.ToArray();
+        }
+    }
+
+    void SetSelectedPass(int index, bool init = false)
+    {
+        if (init || index != m_SelectedPass)
+        {
+            m_SelectedPass = index;
+
+            m_SelectedKeywords.Clear();
+
+            if (m_SelectedPass >= 0 && m_SelectedPass < m_Passes.Count)
+            {
+                PassInfo info = m_Passes[m_SelectedPass];
+
+                for (int i = 0; i < info.keywords.Count; i++)
+                {
+                    m_SelectedKeywords.Add(0);
+                }
+            }
+        }
+    }
+
+    void ParseKeywords(string code, ref List<List<string>> keywords)
+    {
+        keywords.Clear();
+
+        int pragmaIndex = code.IndexOf("#pragma");
```

Comment:
I tested this with the `Standard` shader and there's this in there:
```
            // Uncomment the following line to enable dithering LOD crossfade. Note: there are more in the file to uncomment for other passes.
            //#pragma multi_compile _ LOD_FADE_CROSSFADE
```
This is not a trivial problem to solve, but the commented line is actually picked up by this keyword processing

Reply:
> I see some problems with the keyword detection code:
> 
> * `#include` statements are not executed so a lot of shader code is actually missing
> * Comments (both types, also trailing comments) are not handled resulting in too many keywords
> * `#ifdef` shader preprocessor macros might enable/disable shader keywords? Not 100% sure on this one.
> 
> I didn't review everything after I noticed this because it might require a number of changes.

There shouldn't be any includes in the compiled glsl as they should be 'included' in the compiling process so any extra pragmas included should be there in the single source file.

Reply:
Not sure if the #ifdef is a valid concern, I mean if someone is trying to do that it feels 'daft' much better to combine defines to disable functionality than to disable functionality by attempting to remove a pragma.

Also not sure if its even allowed, if the pre-processor follows the c standard then:

> The standard is unclear on where a _Pragma operator can appear. The preprocessor does not accept it within a preprocessing conditional directive like ‘#if’. To be safe, you are probably best keeping it out of directives other than ‘#define’, and putting it on a line of its own.
> 
> This manual documents the pragmas which are meaningful to the preprocessor itself. Other pragmas are meaningful to the C or C++ compilers. They are documented in the GCC manual.

Reply:
I just learned today about this package that takes care of a lot of that for you: https://github.com/needle-tools/shader-variant-explorer

It relies on the “Preprocess only” option for shaders.